libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt

```diff
@@ -1074,64 +1074,91 @@ internal object MapboxRouteLineUtils {
         return expressionBuilder.build()
     }
 
-    private fun addSource(
-        style: Style,
-        layerSource: String,
+    /**
+     * Compares provided [properties] map with existing values for a source with [sourceId].
+     *
+     * The primary mechanism to compare the values is [Value.equals],
+     * however, there's a special case [Double]s which are compared with a delta.
+     * Delta comparison is not supported for collections
+     * but it's a case that this class currently doesn't need to take into account.
+     */
+    private fun Style.sourcePropertiesCompatible(
+        sourceId: String,
+        properties: HashMap<String, Value>
+    ) = properties.all {
+        val existing = getStyleSourceProperty(sourceId, it.key).value
+        val existingContents = existing.contents
+        val expected = it.value
+        val expectedContents = expected.contents
+        if (existingContents is Double && expectedContents is Double) {
+            abs(existingContents - expectedContents) < 0.000001
+        } else {
+            existing == expected
+        }
+    }
+
+    private fun Style.addNewOrReuseSource(
+        id: String,
         tolerance: Double,
         useLineMetrics: Boolean,
         enableSharedCache: Boolean
     ) {
-        if (!style.styleSourceExists(layerSource)) {
-            style.addStyleSource(
-                layerSource,
+        val source = if (styleSourceExists(id)) {
+            getSourceAs<GeoJsonSource>(id)
+        } else {
+            null
+        }
+
+        val expectedProperties = hashMapOf(
+            "type" to Value("geojson"),
+            "sharedCache" to Value(enableSharedCache),
+            "maxzoom" to Value(16),
+            "lineMetrics" to Value(useLineMetrics),
+            "tolerance" to Value(tolerance),
+        )
+        val recreateSource = source == null || !sourcePropertiesCompatible(id, expectedProperties)
+        if (recreateSource) {
+            source?.let {
+                removeStyleSource(it.sourceId)
+            }
+            addStyleSource(
+                id,
                 Value(
-                    hashMapOf(
-                        "type" to Value("geojson"),
-                        "data" to Value(""),
-                        "sharedCache" to Value(enableSharedCache),
-                        "maxzoom" to Value(16),
-                        "lineMetrics" to Value(useLineMetrics),
-                        "tolerance" to Value(tolerance)
-                    )
+                    expectedProperties.apply {
+                        // reuse feature collections when re-adding the source
+                        put("data", Value(source?.data ?: ""))
```

Comment:
So you compare everything except for the feature collection. So the properties will be compatible if we add a source for a different route, right?